### PR TITLE
Typing fixes to napari.layers.utils

### PR DESCRIPTION
**Pull request by dstansby**
# Fixes/Closes

More work towards https://github.com/napari/napari/issues/2751

# Description
This fixes typing in `napari.layers.utils.*`. There's quite a lot of non-trivial fixes that need to be made (e.g. around `Array` and colour value typing) that I've left, but this fixes all the easy to fix stuff and ticks a few more files of the typing ignore list.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [x] Maintenance (changes required to run napari, tests, & CI smoothly)

### CODE REVIEW

File: napari/layers/utils/_link_layers.py [hunk 1]
```diff
@@ -35,9 +37,10 @@ def get_linked_layers(*layers: Layer) -> Set[Layer]:
     directly linked to each other.  This is useful for context menu generation.
     """
     if not layers:
-        return {}
+        return set()
```
Comment (Carreau) [inline napari/layers/utils/_link_layers.py:40]: +1 good catch.

File: napari/layers/utils/_link_layers.py [hunk 2]
```diff
@@ -35,9 +37,10 @@ def get_linked_layers(*layers: Layer) -> Set[Layer]:
...
     refs = set.union(*(_LINKED_LAYERS.get(ref(x), set()) for x in layers))
-    return {x() for x in refs if x() is not None}
+    linked_layers = {x() for x in refs}
+    return {x for x in linked_layers if x is not None}
```
Comment (Carreau) [inline napari/layers/utils/_link_layers.py:43]: Good catch as well, I guess x can be gc'collected between the `if x()` and the `{x()` !

File: napari/layers/utils/plane.py [hunk 3]
```diff
@@ -35,10 +35,6 @@ def _normalise_vector(cls, v):
     def _ensure_tuple(cls, v):
         return tuple(v)
 
-    def shift_along_normal_vector(self, distance: float):
-        """Shift the plane along its normal vector by a given distance."""
-        self.position += distance * self.normal
-
```
Comment (Carreau) [inline napari/layers/utils/plane.py:41]: I guess that was public, I guess even if unused we might want a deprecation ... @napari/core-devs ?

**Review by Carreau — APPROVED** [review]
Thanks you for all your updates, I hope they get merged soon.

**Comment by Carreau** [comment]
There seem to be an actual type error to fix as well.

File: napari/layers/utils/layer_utils.py [hunk 4]
```diff
@@ -809,7 +819,7 @@ def currents(self) -> Dict[str, np.ndarray]:
 
     def set_currents(
         self,
-        currents: Dict[str, np.ndarray],
+        currents: Dict[str, npt.NDArray],
```
Comment (andy-sweet) [inline napari/layers/utils/layer_utils.py:822]: To clarify, `npt.NDArray` is slightly stronger than `np.ndarray`? Specifically that the `dtype` is a scalar?

File: napari/layers/utils/plane.py (hunk 3 above)
Comment (andy-sweet) [reply napari/layers/utils/plane.py:41]: This class is also exposed through the `Image.plane` property (and maybe others), but I also see it's generally broken right now. I guess if you pass through a positive integer distance, the code will not fail, but it also won't do what's expected. Maybe we could just fix it by iterating over the elements and multiplying as I assume is intended.

@brisvag : what are your thoughts about this?

File: napari/layers/utils/interaction_box.py [hunk 5]
```diff
@@ -77,7 +77,7 @@ def generate_transform_box_from_layer(
     np.ndarray
         Vertices and handles of the interaction box in data coordinates.
     """
-    bounds = layer._display_bounding_box_augmented(dims_displayed)
+    bounds = layer._display_bounding_box_augmented(np.asarray(dims_displayed))
```
Comment (andy-sweet) [inline napari/layers/utils/interaction_box.py]: Why not `list(...)` here? That seems to be what `Layer. _display_bounding_box_augmented` wants?

**Review by andy-sweet — COMMENTED** [review]
Thanks for the fixes. Generally looks good. Will approve after [the removal/deprecation discussion](https://github.com/napari/napari/pull/6020/files#r1254195300) is resolved.

File: napari/layers/utils/_link_layers.py (hunk 2 above)
Comment (brisvag) [reply napari/layers/utils/_link_layers.py:43]: Alternatively:

```
    return {layer for x in refs if (layer := x()) is not None}
```

File: napari/layers/utils/plane.py (hunk 3 above)
Comment (brisvag) [reply napari/layers/utils/plane.py:41]: If this is used at all, it's probably by our plane example or maybe by @alisterburt in napari-threedee. But why does it need removal? I think this was left as a utility for plugins and users.

File: napari/layers/utils/interactivity_utils.py [hunk 6]
```diff
@@ -116,15 +116,17 @@ def orient_plane_normal_around_cursor(layer: Image, plane_normal: tuple):
...
     # early exit
     if viewer.dims.ndisplay != 3 or layer.depiction != VolumeDepiction.PLANE:
         return
 
     # find cursor-plane intersection in data coordinates
     cursor_position = layer._world_to_displayed_data(
-        position=viewer.cursor.position,
-        dims_displayed=layer._slice_input.displayed,
+        position=np.asarray(viewer.cursor.position),
+        dims_displayed=np.asarray(layer._slice_input.displayed),
```
Comment (Carreau) [inline napari/layers/utils/interactivity_utils.py]: I believe this should stay a list as well:

```
 def _world_to_displayed_data(
        self, position: np.ndarray, dims_displayed: List[int]
    ) -> npt.NDArray:
```

File: napari/layers/utils/interaction_box.py (hunk 5 above)
Comment (Carreau) [reply napari/layers/utils/interaction_box.py]:
```suggestion
    bounds = layer._display_bounding_box_augmented(list(dims_displayed))
```

File: napari/layers/utils/interactivity_utils.py (hunk 6 above)
Comment (Carreau) [reply napari/layers/utils/interactivity_utils.py]: And actually 

```
   37     @property
   38     def displayed(self) -> List[int]:
   39         """The layer dimension indices displayed in this slice."""
   40         return list(self.order[-self.ndisplay :])
```

so the original should have been fine.

```suggestion
        dims_displayed=layer._slice_input.displayed,
```

**Comment by Carreau** [comment]
If you don't mind I pushed a commit that adresses @andy-sweet review and other mypy errors from CI. Mostly to see if CI passes, or if other errors get triggered.

Feel free to revert it or amend it.

File: napari/layers/utils/plane.py (hunk 3 above)
Comment (andy-sweet) [reply napari/layers/utils/plane.py:41]: Doesn't need to be removed, but it does need to be fixed if we keep it. I implemented a quick fix as a PR to this one: https://github.com/dstansby/napari/pull/2

Comment (dstansby) [reply napari/layers/utils/plane.py:41]: Thanks, I have no memory of why I removed it... merged your PR to my fork, so should be good now.

**Comment by dstansby** [comment]
As far as I can tell all the comments are addressed now (thanks to those who added their commits to this :smile: ), and I just need to fix the current typing errors.

**Review by andy-sweet — APPROVED** [review]
Looks good now. Thanks again!

**Comment by GenevieveBuckley** [comment]
Two CircleCI docs build CI checks have failed. I've just re-started the CircleCI workflows to see if it's a repeated problem, or just being flaky.

**Pull request merged**